zipline/pipeline/graph.py

```diff
@@ -207,53 +221,69 @@ class ExecutionPlan(TermGraph):
 
     Parameters
     ----------
+    domain : zipline.pipeline.domain.Domain
+        The domain of execution for which we need to build a plan.
     terms : dict
         A dict mapping names to final output terms.
-    all_dates : pd.DatetimeIndex
-        An index of all known trading days for which ``terms`` will be
-        computed.
     start_date : pd.Timestamp
         The first date for which output is requested for ``terms``.
     end_date : pd.Timestamp
         The last date for which output is requested for ``terms``.
 
     Attributes
     ----------
+    domain
+    extra_rows
     outputs
     offset
-    extra_rows
 
     Methods
     -------
-    ordered()
+    ordered(self)
         Return a topologically-sorted iterator over the terms in self.
+    execution_order(self, refcounts)
+        Return a topologically-sorted iterator over the terms in self, skipping
+        entries with refcounts of zero.
     """
     def __init__(self,
+                 domain,
                  terms,
-                 all_dates,
                  start_date,
                  end_date,
                  min_extra_rows=0):
         super(ExecutionPlan, self).__init__(terms)
 
+        # Specialize loadable terms.
+        # NOTE: This isn't using self.loadable_terms because we don't want to
+        # trigger caching of the lazyval.
+        specializations = {
+            t: t.specialize(domain)
+            for t in self.graph if isinstance(t, LoadableTerm)
+        }
+        self.graph = nx.relabel_nodes(self.graph, specializations)
+        self.domain = domain
+
+        sessions = domain.all_sessions()
         for term in terms.values():
             self.set_extra_rows(
                 term,
-                all_dates,
+                sessions,
                 start_date,
                 end_date,
                 min_extra_rows=min_extra_rows,
             )
 
+        self._assert_all_loadable_terms_specialized_to(domain)
+
     def set_extra_rows(self,
                        term,
                        all_dates,
                        start_date,
                        end_date,
                        min_extra_rows):
-        """
-        Compute ``extra_rows`` for transitive dependencies of ``root_terms``
-        """
+        # Specialize any loadable terms before adding extra rows.
+        term = maybe_specialize(term, self.domain)
```

Comment:
I call specialize in a lot of distinct places. Could we lift this to a single location?

Reply:
We've talked about this a bunch in person. The general pattern in this implementation is that anything that comes out of `TermGraph` should be correctly specialized already, but anything that's being fetched from a term (in particular, `term.inputs` and `term.dependencies`) needs to be wrapped in a `maybe_specialize` call.

I think we agree that the right long term design for this is to specialize the entire term graph in the place where we currently specialize loadable terms. Implementing that either requires adding custom `specialize` implementations to 10 or so classes, or it requires refactoring `Term` a fair amount to allow reliably making functional copies of any term. I think I see a path toward the second option, but I'd prefer to push that into a follow-up, since I'm fairly confident that this implementation is correct, and I don't want to block downstream work on this.